pkg/storage/chunk/local/boltdb_index_client.go

```diff
@@ -282,16 +290,33 @@ func (b *BoltIndexClient) QueryWithCursor(_ context.Context, c *bbolt.Cursor, qu
 			continue
 		}
 
+		// we need to do callback only once to pass the batch iterator
+		if !callbackDone {
+			wg.Add(1)
+			// do the callback in a goroutine to stream back the index entries
+			go func() {
+				// wait for callback to finish processing the batch and return
+				defer wg.Done()
+				callback(query, batch)
+			}()
+			callbackDone = true
+		}
+
 		// make a copy since k, v are only valid for the life of the transaction.
 		// See: https://godoc.org/github.com/boltdb/bolt#Cursor.Seek
-		batch.rangeValue = make([]byte, len(k)-len(rowPrefix))
-		copy(batch.rangeValue, k[len(rowPrefix):])
+		rangeValue := make([]byte, len(k)-len(rowPrefix))
+		copy(rangeValue, k[len(rowPrefix):])
 
-		batch.value = make([]byte, len(v))
-		copy(batch.value, v)
+		value := make([]byte, len(v))
+		copy(value, v)
 
-		if !callback(query, &batch) {
-			break
+		err := batch.send(singleResponse{
+			rangeValue: rangeValue,
+			value:      value,
+		})
+
+		if err != nil {
+			return err
```

Comment:
may be worth wrapping this error? Because from the `send` method(), the error can be simply "timedout sending response". which I believe lacks context when I see this error in practice.